pkg/envoy/accesslog/request_header_operator.go

```diff
@@ -0,0 +1,84 @@
+package accesslog
+
+import (
+	"fmt"
+
+	accesslog_config "github.com/envoyproxy/go-control-plane/envoy/config/accesslog/v2"
+	accesslog_data "github.com/envoyproxy/go-control-plane/envoy/data/accesslog/v2"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+)
+
+// RequestHeaderOperator represents a `%REQ(X?Y):Z%` command operator.
+type RequestHeaderOperator struct {
+	HeaderFormatter
+}
+
+func (f *RequestHeaderOperator) FormatHttpLogEntry(entry *accesslog_data.HTTPAccessLogEntry) (string, error) {
+	return f.Format(&RequestHeaders{entry.GetRequest()})
+}
+
+func (f *RequestHeaderOperator) FormatTcpLogEntry(entry *accesslog_data.TCPAccessLogEntry) (string, error) {
+	return "", nil
+}
+
+func (f *RequestHeaderOperator) ConfigureHttpLog(config *accesslog_config.HttpGrpcAccessLogConfig) error {
+	config.AdditionalRequestHeadersToLog = f.AppendTo(config.AdditionalRequestHeadersToLog)
+	return nil
+}
+
+func (f *RequestHeaderOperator) ConfigureTcpLog(config *accesslog_config.TcpGrpcAccessLogConfig) error {
+	// has no effect on TcpGrpcAccessLogConfig
+	return nil
+}
+
+// String returns the canonical representation of this access log fragment.
+func (f *RequestHeaderOperator) String() string {
+	return fmt.Sprintf("%%REQ%s%%", f.HeaderFormatter.String())
+}
+
+// RequestHeaders represents a set of HTTP request headers
+// that includes both regular headers, such as `referer` and `user-agent`,
+// and pseudo headers, such as `:method`, `:authority` and `:path`.
+type RequestHeaders struct {
+	*accesslog_data.HTTPRequestProperties
+}
+
+func (h *RequestHeaders) Get(name string) (string, bool) {
+	switch name {
+	case ":method":
+		return h.formatHttpMethod(h.GetRequestMethod())
+	case ":scheme":
+		return h.optionalString(h.GetScheme())
+	case ":authority":
+		return h.optionalString(h.GetAuthority())
+	case ":path":
+		return h.optionalString(h.GetPath())
+	case "user-agent":
+		return h.optionalString(h.GetUserAgent())
+	case "referer":
+		return h.optionalString(h.GetReferer())
+	case "x-forwarded-for":
+		return h.optionalString(h.GetForwardedFor())
+	case "x-request-id":
+		return h.optionalString(h.GetRequestId())
+	case "x-envoy-original-path":
+		return h.optionalString(h.GetOriginalPath())
+	default:
+		value, exists := h.GetRequestHeaders()[name]
```

Comment:
Envoy represents request headers as `string-to-string` entries.

So, the way Envoy handles request headers with multiple values is not specific to Access Log. 

In practice, I observe the following behaviour:
* in case of a known header, such as `Referer`, Envoy combines multiple values into a single string, e.g.
  ```
  curl \
  -H'Referer: google.com' \
  -H'Referer: bing.com' \
  localhost:3003/ 
  ```
  causes Envoy to emit the following log entry
  ```
    request:
      authority: localhost:3003
      path: /
      referer: google.com,bing.com # notice that Envoy combined multiple values into a single string
      requestHeadersBytes: "247"
      requestId: 53e6366e-e85a-4d98-acb9-e8495b13b825
      requestMethod: GET
      scheme: http
      userAgent: curl/7.64.1
  ```
* in case of an unknown header, such as `X-My-Header`, Envoy uses only the first value
  ```
  curl \
  -H'x-my-header: value-2' \
  -H'x-my-header: value-1' \
  localhost:3003/ 
  ```
  causes Envoy to emit the following log entry
  ```
      request:
      authority: localhost:3003
      path: /
      requestHeaders:
        x-my-header: value-2 # notice that only the first header value is preserved
      requestHeadersBytes: "247"
      requestId: 53e6366e-e85a-4d98-acb9-e8495b13b825
      requestMethod: GET
      scheme: http
      userAgent: curl/7.64.1
  ```

HTTP/1.1 spec [says the following](https://tools.ietf.org/html/rfc2616#page-32) about this situation:

```
   Multiple message-header fields with the same field-name MAY be
   present in a message if and only if the entire field-value for that
   header field is defined as a comma-separated list [i.e., #(values)].
   It MUST be possible to combine the multiple header fields into one
   "field-name: field-value" pair, without changing the semantics of the
   message, by appending each subsequent field-value to the first, each
   separated by a comma. The order in which header fields with the same
   field-name are received is therefore significant to the
   interpretation of the combined field value, and thus a proxy MUST NOT
   change the order of these field values when a message is forwarded.
```

Reply:
Thanks for looking into it